docs: Block content proof design (#873)
hiero-ledger/hiero-block-node: block-node/server/docs/design/block-content-proof.md

### PBJBlockContentProofServiceProxy

PBJBlockContentProofServiceProxy is the entity responsible for handling the block content proof requests. It provides the implementation for the `getBlockContentProof` rpc endpoint.
Handles the verification of webServerStatus, handles exceptions, and transforms the response from the BlockContentProofService to a `BlockContentProofResponse` message.

### BlockContentProofService

A java module that embodies the functionality necessary to respond to request for block content proofs by calculating and returning those proofs

### BlockProvider

Shall provide a block to the BlockContentProofService at their request. `BlockContentProofService` will request a Block using a message request, and the `BlockProvider` will return the block.

### BlockMerkleTreeInfo

Is a record type that contains all the necessary data to compute any block item proof in the block.

## Design

1. The `PBJBlockContentProofServiceProxy` receives the block content proof requests.
2. If the request is valid and the web server is running, it forwards the request to the `BlockContentProofService`.
3. The `BlockContentProofService` publish a message requests for the block.

> 4. The `BlockProvider` provides the Block to the `BlockContentProofService`. If block is Available.

Reviewer: Do we want to note that this component might be a separate facility or service, and is outside the scope of this design document?

5. `BlockContentProofService` uses the provided block to calculate a `BlockContentProof`.
6. The `BlockContentProofService` returns the `BlockContentProof` to the `PBJBlockContentProofServiceProxy`.
7. The `PBJBlockContentProofServiceProxy` wraps the `BlockContentProof` in a `BlockContentProofResponse` and sends it back to the client.

## Sequence Diagram

```mermaid
sequenceDiagram
    participant Client as Client
    participant Proxy as PBJBlockContentProofServiceProxy
    participant Service as BlockContentProofService

>     participant Reader as BlockReader

Reviewer: Should this be `BlockProvider`?

Author: Yes, good catch, should be fixed

> # Design Document Template

Reviewer 2: ```suggestion
# Block Contents Proof Design Document
```

Author: Good Catch!! Thank you!

## Table of Contents

1. [Purpose](#purpose)
2. [Goals](#goals)
3. [Terms](#terms)
4. [Entities](#entities)
5. [Design](#design)
6. [Sequence Diagram](#sequence-diagram)
7. [Configuration](#configuration)
8. [Metrics](#metrics)
9. [Exceptions](#exceptions)
10. [Acceptance Tests](#acceptance-tests)

## Purpose

> The purpose of this document is to define the design of the BlockContentProof feature.

Reviewer 2: nit: we should standardize but initially i thought it was block contents.
Here I see Block Content.
I block has contents but I'm guessing the singular form is because I since content items (BlockItem) is under consideration.

The BlockContentProof API allows the BN to provide a proof of the content of a block to the client.

## Goals

> 1. Define the BlockContentProof struct

Reviewer 2: By struct to did you mean structure? Logic might be better.

Author: yes

2. Define the BlockContentProof API

> 3. BN should be able to calculate the BlockContentProof given a BlockNumber and BlockItem (or BlockItemHash or BlockItemIndex)

Reviewer 2: ```suggestion
3. Given a BlockNumber and BlockItem (or BlockItemHash or BlockItemIndex) a BN should support the calculation of a BlockContentProof
```

4. A Client should be able to verify the BlockContentProof easily.

## Terms

<dl>
  <dt>BlockNumber</dt>
  <dd>A value that represents the unique number (identifier) of a given `Block`.
  It is a strictly increasing `long` value, starting from zero (`0`).</dd>

  <dt>BlockItem</dt>
  <dd>Every block is comprised of a repeated set of block_items, there are many types of block_items, but for the purpose of understanding the BlockContentProof we can divide them into 2 large groups, `inputs` and `outputs` each Kind of BlockItem can belong to these groups and this is very important in order to calculate correctly the BlockRootHash and a BlockContentProof. </dd>

  <dt>BlockItemHash</dt>
  <dd>Is the SHA-384 hash of the block_item </dd>

  <dt>MerkleSiblingHashes</dt>
  <dd>The ordered list of hashes for each "sibling" in the binary tree from the item to be proven all the way up to the tree root.</dd>

  <dt>BlockContentProof</dt>

>   <dd>A data structure that holds all the needed information to prove a given block item is part of a given block number. Is at least comprised of `BlockNumber, `BlockItem`, `MerkleSiblingHashes`, `BlockSignature`</dd>

Reviewer 2: ```suggestion
  <dd>A data structure that holds all the needed information to prove a given block item is part of a given block number. Is at least comprised of the `BlockItem`, `MerkleSiblingHashes`, `BlockSignature`</dd>
```

# Design Document Template

## Table of Contents

1. [Purpose](#purpose)
2. [Goals](#goals)
3. [Terms](#terms)
4. [Entities](#entities)
5. [Design](#design)
6. [Sequence Diagram](#sequence-diagram)
7. [Configuration](#configuration)
8. [Metrics](#metrics)
9. [Exceptions](#exceptions)
10. [Acceptance Tests](#acceptance-tests)

## Purpose

The purpose of this document is to define the design of the BlockContentProof feature.
The BlockContentProof API allows the BN to provide a proof of the content of a block to the client.

## Goals

1. Define the BlockContentProof struct
2. Define the BlockContentProof API
3. BN should be able to calculate the BlockContentProof given a BlockNumber and BlockItem (or BlockItemHash or BlockItemIndex)
4. A Client should be able to verify the BlockContentProof easily.

## Terms

<dl>
  <dt>BlockNumber</dt>
  <dd>A value that represents the unique number (identifier) of a given `Block`.
  It is a strictly increasing `long` value, starting from zero (`0`).</dd>

  <dt>BlockItem</dt>

>   <dd>Every block is comprised of a repeated set of block_items, there are many types of block_items, but for the purpose of understanding the BlockContentProof we can divide them into 2 large groups, `inputs` and `outputs` each Kind of BlockItem can belong to these groups and this is very important in order to calculate correctly the BlockRootHash and a BlockContentProof. </dd>

Reviewer: I wonder if we need this detail here.  The exact mechanics of computing a block hash are defined in the Block Stream HIP, and _should not_ be relevant here except that the block item is placed in the _correct location_ within the Block Merkle tree (wherever the HIP defines that to be).

Author: Improved wording 👍 Thank you!!

Reviewer: This did not change...

  <dt>BlockItemHash</dt>
  <dd>Is the SHA-384 hash of the block_item </dd>

  <dt>MerkleSiblingHashes</dt>

>   <dd>Are the sibling missing hashes from the item all the way up to the BlockRootHash.</dd>

Reviewer: These shouldn't be _missing_.
```suggestion
  <dd>The ordered list of hashes for each "sibling" in the binary tree from the item to be proven all the way up to the tree root.</dd>
```

  <dt>BlockContentProof</dt>

>   <dd>A data structure that holds all the needed information to prove a given block item is part of a given block number. Is at least comprised of `BlockNumber, `BlockItemHash`, `MerkleSiblingHashes`, `BlockSignature`</dd>

Reviewer: I wonder if this should be closer to what we return:
```suggestion
  <dd>A data structure that holds all the needed information to prove a given block item is part of a given block number. Is at least comprised of `BlockNumber, `BlockItem`, `MerkleSiblingHashes`, `BlockSignature`</dd>
```

  <dt>BlockRootHash</dt>

>   <dd>Is a "special" merkle tree defined by the `Hiero` DLT that is comprised by `previousBlockHash`, `inputsMerkleTree`, `outputsMerkleTree` and `stateRootHash`.</dd>

Reviewer: This is defined in the Block Streams HIP, not by the DLT, and we shouldn't specify it here (as it's subject to change out of sequence with this document).
```suggestion
  <dd>The SHA2-384 value at the root of a carefully defined, partially truncated, binary merkle tree defined by the Block Streams HIP (1056).</dd>
```

  <dt>Block Signature</dt>

>   <dd>Is the Signed BlockRootHash by the Network, its purpose is to verify that the BlockRootHash of a given block was signed by the network.</dd>

Reviewer: The wording here is a bit awkward.
```suggestion
  <dd>A TSS signature of a BlockRootHash provided by the consensus network in a `BlockProof` item. Its purpose is to verify that the full content of a given block was generated by and agreed upon by the consensus network as a whole.</dd>
```

</dl>

## API

### block_proof.proto

```protobuf
message MerkleSiblingHash {
    bool is_first = 1;
    bytes sibling_hash = 2;
}

> ```

Reviewer: Not major, but we usually don't want to re-specify a pre-existing protocol buffer message.

Author: just removed the message, since adding a link to the whole API spec, they can find it if needed.

### block_proof_service.proto

```protobuf
 message BlockContentProof {
  com.hedera.hapi.block.stream.BlockItem block_item = 1;
  repeated MerkleSiblingHash sibling_hashes = 2;
  bytes block_signature = 3;
}

message BlockContentProofRequest {
    uint64 block_number = 1;
    oneof block_item {
        com.hedera.hapi.block.stream.BlockItem block_item = 1;
        bytes block_item_hash = 2;
        uint32 block_item_index = 3;
    }
}

enum BlockContentProofResponseCode {
    BLOCK_CONTENT_PROOF_UNKNOWN = 0;
    BLOCK_CONTENT_PROOF_SUCCESS = 1;

>     BLOCK_CONTENT_PROOF_NOT_FOUND = 2;

Reviewer: What does not found mean?  The _item_ was not found, or the _block_ was not found?
I am assuming it means the _item_ was not found.

Reviewer 2: I think `BLOCK_CONTENT_PROOF_NOT_FOUND` is still unclear no?
Only the block or block items exist to be found. The proof is based on those.
So the not found should refer to one or both of those

Reviewer: That would probably make this clearer.
```suggestion
    BLOCK_CONTENT_PROOF_ITEM_NOT_FOUND = 2;
```

>     BLOCK_CONTENT_PROOF_INSUFFICIENT_BALANCE = 3;

Reviewer: We don't even know if there will be any monetization or charging, and we certainly won't be checking balance in this service.
We do need a response for "not available" (e.g. could not get the block, or could not compute the proof).
```suggestion
    BLOCK_CONTENT_PROOF_NOT_AVAILABLE = 3;
```
This would replace `INTERNAL_ERROR` below as well.

    BLOCK_CONTENT_PROOF_INTERNAL_ERROR = 4;
    BLOCK_CONTENT_PROOF_DUPLICATE_HASH_ITEM_FOUND = 5;
}

message BlockContentProofResponse {
    BlockContentProofResponseCode status = 1;
    com.hedera.hapi.block.stream.BlockContentProof block_content_proof = 2;
}

service BlockContentProofService {
    rpc getBlockContentProof(BlockContentProofRequest) returns (BlockContentProofResponse);

> }

Reviewer: For new messages, should we include the full specification?

Author: I added a Link to the full spec API, I think that due to the comments, it makes it too large. I wonder if we should not remove the API section altogether once we merge the API PR, and leave only the references here.

Reviewer: Just references makes sense, to me, for an in-tree design document.

```

## Entities

### BlockContentProof

> BlockContentProof is a new record that contains all the necessary data to compute the block content proof. It contains the following fields:

Reviewer: This isn't all of the necessary data, as we still need the block itself (including block proof item).  Should we add that value here, or reword this sentence?
```suggestion
BlockContentProof is a new record that contains the fields necessary to compute the block content proof when combined with the retrieved block data. It contains the following fields:
```

- block_number: long,
- block_item: BlockItem,
- block_item_hash: Bytes,
- sibling_hashes: List< MerkleSiblingHash>,
- block_root_hash: Bytes,
- block_signature: Bytes

### PBJBlockContentProofServiceProxy

PBJBlockContentProofServiceProxy is the entity responsible for handling the block content proof requests. It provides the implementation for the `getBlockContentProof` rpc endpoint.

> Handles the verification of webServerStatus, handles exceptions, and wraps the response from the BlockContentProofService in a `BlockContentProofResponse` message.

Reviewer: We aren't wrapping the value so much as transforming it.
```suggestion
Handles the verification of webServerStatus, handles exceptions, and transforms the response from the BlockContentProofService to a `BlockContentProofResponse` message.
```

### BlockContentProofService

> Requests the Block to the BlockReader (Persistence Module), waits to get it and once it gets the Block it creates a new BlockMerkleTreeInfo record out of that block, after that it uses it to create the BlockContentProof and returns it to the PBJBlockContentProofServiceProxy.

Reviewer: This is very confusing what the service entity is intended to represent. I'm entirely uncertain whether this is supposed to be the whole service, or a class that has the same name.

This is a guess that the intent was to describe the _service_ and not an implementation class, but perhaps the following is clearer?
```suggestion
A java module that embodies the functionality necessary to respond to request for block content proofs by calculating and returning those proofs.
```

### BlockReader

> The BlockReader is responsible for reading the block items from the block store and providing them to the BlockContentProofService.

Reviewer: We do _not_ want to directly interact with block readers or the block data store here, as there is no guarantee it exists in any particular form (it could be local, remote, stored in a columnar database, or even requested from other block nodes on request).

Perhaps refer to the "Persistence Service" and describe a block read message/response sequence to obtain the necessary block instead?

### BlockMerkleTreeInfo

Is a new record type that contains all the necessary data to compute any block item proof in the block. it contains:
- inputsMerkleTree: List< List< Bytes>>,
- outputsMerkleTree: List< List< Bytes>>,
- previousBlockHash: Bytes,
- stateRootHash: Bytes,

> - blockHash: Bytes,

Author: removed!

2. If the request is valid and the web server is running, it forwards the request to the `BlockContentProofService`.
3. The `BlockContentProofService` requests the block from the `BlockReader`.
4. The `BlockReader` reads the block items from the block store and provides them to the `BlockContentProofService`.
5. `BlockContentProofService` calculates the `BlockMerkleTreeInfo` and creates the `BlockContentProof` using the `BlockMerkleTreeInfo`.
6. The `BlockContentProofService` returns the `BlockContentProof` to the `PBJBlockContentProofServiceProxy`.
7. The `PBJBlockContentProofServiceProxy` wraps the `BlockContentProof` in a `BlockContentProofResponse` and sends it back to the client.

## Sequence Diagram

```mermaid
sequenceDiagram
    participant Client as Client
    participant Proxy as PBJBlockContentProofServiceProxy
    participant Service as BlockContentProofService
    participant Reader as BlockReader

    Client->>Proxy:  BlockContentProofRequest
    Proxy->>Proxy:  Validates request & check web server status
    Proxy->>Service: BlockContentProofRequest
    Service->>Reader: Requests Block Number

>     Reader-->>Service: Returns Block items

Reviewer: As we're dealing with already-complete blocks in this case, the return from whatever service manages stored blocks should be a `Block`, not the individual items.
```suggestion
    Reader-->>Service: Returns Block
```

    Service->>Service: Calculate BlockMerkleTreeInfo
    Service->>Service: Create BlockContentProof using BlockMerkleTreeInfo
    Service-->>Proxy: Return BlockContentProof
    Proxy->>Client: BlockContentProofResponse

```

## Configuration

## Metrics

1. BlockContentProofRequest_Success Counter
2. BlockContentProofRequest_Error Counter
3. BlockContentProofResponse_Latency Histogram

**Note:** Once we are able to add labels we should add `status` label to the metrics

## Exceptions

> Any exception will be handled and return `BLOCK_CONTENT_PROOF_INTERNAL_ERROR` response code to the client.

Reviewer: As Jasper noted, we shouldn't return internal errors:
```suggestion
Any exception will be handled and return `BLOCK_CONTENT_PROOF_NOT_AVAILABLE` response code to the client.
```

3. Client receives the BlockContentProofResponse and verifies the block_item using the BlockContentProof provided by the BN.

### Edge Case duplicate block_item (Failure Case):

1. Client sends a BlockContentProofRequest using `block_item_hash` to the BN for a block_item that has a duplicate hash in the block.
2. BN receives the request and returns a `BLOCK_CONTENT_PROOF_DUPLICATE_HASH_ITEM_FOUND` response code to the client.
3. Client receives the response and verify is the correct response code.

### Edge Case duplicate block_item (Success Case):

1. Client sends a BlockContentProofRequest using `block_item_index` to the BN for a block_item that has a duplicate hash in the block.
2. BN receives the request and returns a complete BlockContentProofResponse.
3. Client receives the BlockContentProofResponse and verifies the block_item using the BlockContentProof provided by the BN.
4. Client verifies that the block_item is the correct one by pre-calculating the merkle proof path before hand for the right item and then comparing the given proof with the expected proof.

### Block Number not found:

1. Client sends a BlockContentProofRequest to the BN for a non-existing block number.

> 2. BN receives the request and returns a `BLOCK_CONTENT_PROOF_NOT_FOUND` response code to the client.

Reviewer: If the _block_ is not available we should return NOT_AVAILABLE.
I would reserve NOT_FOUND for not finding the _item_ within the block.
